### PR TITLE
ARROW-15487: [FlightRPC][C++][GLib][Python][R] Implement FlightClient::Close

### PR DESCRIPTION
**Pull request by lidavidm**
Add a method to explicitly close FlightClient in anticipation of implementing alternative transports which may need this, and to provide an interface for things like ARROW-15473. Because this did not exist before, it is implicitly called by the destructor. For gRPC, this is a no-op.

### CODE REVIEW

File: cpp/src/arrow/flight/client.h [hunk 1]
```diff
@@ -323,8 +323,19 @@ class ARROW_FLIGHT_EXPORT FlightClient {
...
+  /// the destructor if not already called, but this gives the
+  /// application no chance to handle errors, so it is recommended to
+  /// explicitly close the client.
+  ///
+  /// \since 8.0.0
+  Status Close();
+
  private:
   FlightClient();
+  Status CheckOpen();
```
Comment (cyb70289) [inline cpp/src/arrow/flight/client.h:338]: Nit: append `const`?

**Review by cyb70289 — APPROVED** [review]
+1

**Comment by cyb70289** [comment]
@kou, will you have a look at GLib related changes?

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**